specification/cognitiveservices/OpenAI.Inference/models/completions/chat_messages.tsp

```diff
@@ -0,0 +1,202 @@
+  @doc("The chat role associated with this message, which is always 'user' for user messages.")
+  role: ChatRole.user;
+
+  @doc("The contents of the user message, with available input types varying by selected model.")
+  content: ChatMessageContent;
+
+  @doc("An optional name for the participant.")
+  name?: string;
+}
+
+@added(ServiceApiVersions.v2023_05_15)
+@doc("A request chat message representing response or action from the assistant.")
+model ChatRequestAssistantMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'assistant' for assistant messages.")
+  role: ChatRole.assistant;
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
+
+  @doc("An optional name for the participant.")
+  name?: string;
+
+  @projectedName("json", "tool_calls")
+  @added(ServiceApiVersions.v2023_12_01_Preview)
+  @doc("""
+  The tool calls that must be resolved and have their outputs appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  toolCalls?: ChatCompletionsToolCall[];
+
+  @added(ServiceApiVersions.v2023_07_01_Preview)
+  @removed(ServiceApiVersions.v2023_12_01_Preview)
+  @projectedName("json", "function_call")
+  @doc("""
+  The function call that must be resolved and have its output appended to subsequent input messages for the chat
+  completions request to resolve as configured.
+  """)
+  functionCall?: FunctionCall;
+}
+
+@added(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A request chat message representing requested output from a configured tool.")
+model ChatRequestToolMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'tool' for tool messages.")
+  role: ChatRole.tool;
+
+  #suppress "@azure-tools/typespec-azure-core/no-nullable" "explicitly nullable in mirrored API"
+  @doc("The content of the message.")
+  content: string | null;
+
+  @projectedName("json", "tool_call_id")
+  @doc("The ID of the tool call resolved by the provided content.")
+  toolCallId: string;
+}
+
+@added(ServiceApiVersions.v2023_07_01_Preview)
+@removed(ServiceApiVersions.v2023_12_01_Preview)
+@doc("A request chat message representing requested output from a configured function.")
+model ChatRequestFunctionMessage extends ChatRequestMessage {
+  @doc("The chat role associated with this message, which is always 'function' for function messages.")
+  role: ChatRole.function;
+
+  @doc("The name of the function that was called to produce output.")
+  name: string;
+
+  @doc("The output of the function as requested by the function call.")
+  content: string;
+}
+
+@added(ServiceApiVersions.v2023_05_15)
+@doc("A representation of a chat message as received in a response.")
+model ChatResponseMessage {
+  @doc("The chat role associated with the message.")
+  role: ChatRole;
```

Comment:
Correct me if I am wrong here, but this seems to **always** be assistant (at least for azure): the `role` model is separate in the response scenario ([line](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1149)) and that enum only contains assistant according to [this](https://github.com/Azure/azure-rest-api-specs/blob/b456e86a719c459fbd0a5b096b37a2211a519e84/specification/cognitiveservices/data-plane/AzureOpenAI/inference/preview/2023-12-01-preview/inference.yaml#L1161)

Reply:
Hmm, need to think about this one more:
- [OpenAI's documentation](https://platform.openai.com/docs/api-reference/chat/object#chat/object-choices) doesn't constrain it
- But [OpenAI's spec](https://github.com/openai/openai-openapi/blob/a7a0b63d1fac3acac9a223d57da1219c8a97c01e/openapi.yaml#L5907C10-L5907C10) does
- A sticky point might be figuring out what to do with the "context" messages for dataSources; are those a "special" kind of request message?